lib/service/dataLink.js

```diff
@@ -0,0 +1,214 @@
+  'onGetCountWithQuery'
+];
+
+class ServiceObject {
+  constructor(serviceObjectName) {
+    this.eventMap = {};
+    this.serviceObjectName = serviceObjectName;
+  }
+
+  register(dataOp, functionToExecute) {
+    if ((typeof dataOp === 'undefined' && dataOp === null) ||
+      dataRegistrationOperators.indexOf(dataOp) < 0) {
+      throw new Error('Operation not permitted');
+    }
+
+    this.eventMap[dataOp] = functionToExecute;
+  }
+
+  unregister(dataOp) {
+    if ((typeof dataOp === 'undefined' && dataOp === null) ||
+      dataRegistrationOperators.indexOf(dataOp) < 0) {
+      throw new Error('Operation not permitted');
+    }
+
+    delete this.eventMap[dataOp];
+  }
+
+  onInsert(functionToExecute) {
+    this.register('onInsert', functionToExecute);
+  }
+
+  onDeleteById(functionToExecute) {
+    this.register('onDeleteById', functionToExecute);
+  }
+
+  onDeleteAll(functionToExecute) {
+    this.register('onDeleteAll', functionToExecute);
+  }
+
+  onDeleteByQuery(functionToExecute) {
+    this.register('onDeleteByQuery', functionToExecute);
+  }
+
+  onUpdate(functionToExecute) {
+    this.register('onUpdate', functionToExecute);
+  }
+
+  onGetById(functionToExecute) {
+    this.register('onGetById', functionToExecute);
+  }
+
+  onGetAll(functionToExecute) {
+    this.register('onGetAll', functionToExecute);
+  }
+
+  onGetByQuery(functionToExecute) {
+    this.register('onGetByQuery', functionToExecute);
+  }
+
+  onGetCount(functionToExecute) {
+    this.register('onGetCount', functionToExecute);
+  }
+
+  onGetCountWithQuery(functionToExecute) {
+    this.register('onGetCountWithQuery', functionToExecute);
+  }
+
+  removeHandler(handler) {
+    this.unregister(handler);
+  }
+
+  resolve(dataOp) {
+    if (!this.eventMap[dataOp]) {
+      return notImplementedHandler;
+    }
+
+    return this.eventMap[dataOp];
+  }
+}
+
+exports.serviceObject = serviceObject = (serviceObjectName) => {
+  if (!registeredServiceObjects[serviceObjectName]) {
+    registeredServiceObjects[serviceObjectName] = new ServiceObject(serviceObjectName);
+  }
+
+  return registeredServiceObjects[serviceObjectName];
+};
+
+exports.process = (task, modules, callback) => {
+  if (task.request.serviceObjectName === null) {
+    const result = task.response;
+    result.body = kinveyErrors.generateKinveyError('NotFound', 'ServiceObject name not found');
+    result.statusCode = result.body.statusCode;
+    delete result.body.statusCode;
+    return callback(task);
+  }
+
+  // Handle a bug in KCS - if query is part of the task, put it in the request
+  if (!task.request.query) {
+    task.request.query = task.query;
+  }
+
+  const serviceObjectToProcess = serviceObject(task.request.serviceObjectName);
+  let dataOp = '';
+  const dataLinkCompletionHandler = kinveyCompletionHandler(task, callback);
+
+  try {
+    task.request.body = JSON.parse(task.request.body);
+  } catch (e) {
+    if (task.request.body && typeof task.request.body !== 'object') {
```

Comment:
The code was previously testing using `?` -- i.e. existence rather than truthiness. With this new code, a body set to any falsey value (such as `false` or `0` or `''`) will not trigger the error. Same is true for the `if` below for `task.request.query`.